mlflow/pytorch/_pytorch_autolog.py

```diff
@@ -0,0 +1,235 @@
+# In case of MlflowLogger, Run management is completely controlled by the class and
+# hence mlflow object needs to be reinstantiated by setting
+# tracking uri, experiment_id and run_id which may lead to a race condition.
+# TODO: Replace __MlflowPLCallback with Pytorch Lightning's built-in MlflowLogger
+# once the above mentioned issues have been addressed
+
+
+@rank_zero_only
+@experimental
+def _autolog(log_every_n_epoch=1):
+    """
+    Enable automatic logging from pytorch to MLflow.
+    Logs loss and any other metrics specified in the fit
+    function, and optimizer data as parameters. Model checkpoints
+    are logged as artifacts and pytorch model is stored under `model` directory.
+
+    MLflow will also log the parameters of the
+    `EarlyStoppingCallback <https://pytorch-lightning.readthedocs.io/en/latest/early_stopping.html>`
+
+    :param log_every_n_epoch: parameter to log metrics once in `n` epoch. By default, metrics
+                       are logged after every epoch.
+    """
+    global every_n_epoch
+    every_n_epoch = log_every_n_epoch
+
+    class __MLflowPLCallback(pl.Callback):
+        """
+        Callback for auto-logging metrics and parameters.
+        """
+
+        def __init__(self):
+            self.early_stopping = False
+
+        def on_epoch_end(self, trainer, pl_module):
+            """
+            Log loss and other metrics values after each epoch
+
+            :param trainer: pytorch lightning trainer instance
+            :param pl_module: pytorch lightning base module
+            """
+            if (pl_module.current_epoch + 1) % every_n_epoch == 0:
+                for key, value in trainer.callback_metrics.items():
+                    try_mlflow_log(
+                        mlflow.log_metric, key, float(value), step=pl_module.current_epoch
+                    )
+
+            for callback in trainer.callbacks:
+                if isinstance(callback, pl.callbacks.early_stopping.EarlyStopping):
+                    self._early_stop_check(callback)
+
+        def on_train_start(self, trainer, pl_module):
+            """
+            Logs Optimizer related metrics when the train begins
+
+            :param trainer: pytorch lightning trainer instance
+            :param pl_module: pytorch lightning base module
+            """
+            try_mlflow_log(mlflow.set_tag, "Mode", "training")
+            try_mlflow_log(mlflow.log_param, "epochs", trainer.max_epochs)
+
+            for callback in trainer.callbacks:
+                if isinstance(callback, pl.callbacks.early_stopping.EarlyStopping):
+                    self.early_stopping = True
+                    self._log_early_stop_params(callback)
+
+            if hasattr(trainer, "optimizers"):
+                for optimizer in trainer.optimizers:
+                    try_mlflow_log(mlflow.log_param, "optimizer_name", type(optimizer).__name__)
+                    optimizer_name = type(optimizer).__name__.lower() + "_optimizer"
```

Comment:
The current approach doesn't work when we use two Adam optimizers. We should address that in follow-ups.